src/MediawikiApi.php

```diff
@@ -66,21 +67,37 @@ public static function newFromApiEndpoint( $apiEndpoint ) {
 	}
 
 	/**
+	 * Create a new MediawikiApi object from a URL to any page in a MediaWiki website.
+	 *
 	 * @since 2.0
+	 * @see https://en.wikipedia.org/wiki/Really_Simple_Discovery
 	 *
 	 * @param string $url e.g. https://en.wikipedia.org OR https://de.wikipedia.org/wiki/Berlin
-	 *
 	 * @return self returns a MediawikiApi instance using the apiEndpoint provided by the RSD
 	 *              file accessible on all Mediawiki pages
-	 *
-	 * @see https://en.wikipedia.org/wiki/Really_Simple_Discovery
+	 * @throws UsageException If the RSD URL could not be found in the page's HTML.
 	 */
 	public static function newFromPage( $url ) {
-		$tempClient = new Client( array( 'headers' => array( 'User-Agent' => 'addwiki-mediawiki-client' ) ) );
-		$pageXml = new SimpleXMLElement( $tempClient->get( $url )->getBody() );
-		$rsdElement = $pageXml->xpath( 'head/link[@type="application/rsd+xml"][@href]' );
-		$rsdXml = new SimpleXMLElement( $tempClient->get( (string) $rsdElement[0]->attributes()['href'] )->getBody() );
-		return self::newFromApiEndpoint( (string) $rsdXml->service->apis->api->attributes()->apiLink );
+		$tempClient = new Client( [ 'headers' => [ 'User-Agent' => 'addwiki-mediawiki-client' ] ] );
+
+		// Get the page HTML and extract the RSD link.
+		$rsdUrl = false;
+		$pageHtml = $tempClient->get( $url )->getBody();
+		$pageDoc = new DOMDocument();
+		$pageDoc->loadHTML( $pageHtml );
+		foreach ( $pageDoc->getElementsByTagName( 'link' ) as $link ) {
+			$type = $link->attributes->getnamedItem( 'type' );
+			if ( $type && $type->nodeValue == 'application/rsd+xml' ) {
+				$rsdUrl = $link->attributes->getnamedItem( 'href' )->nodeValue;
+			}
+		}
```

Comment:
Let's either use something similar to Tpt's suggestion or add a continue to exit the foreach when an rsd link is found. Otherwise, this could take a while to execute.